bookend: Add stream icons to subscribed/unsubscribed divider line.

akarsh-jain-790 opened the pull request:
Fixes: #23685

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="882" alt="Screenshot 2023-04-15 at 4 11 45 PM" src="https://user-images.githubusercontent.com/72064462/232216507-062c868a-e86a-439d-89bd-b3a532f936a8.png">

<img width="870" alt="Screenshot 2023-04-15 at 4 07 04 PM" src="https://user-images.githubusercontent.com/72064462/232216514-e0401bc8-d602-4ade-aa9d-d89404246acb.png">

<img width="878" alt="Screenshot 2023-04-15 at 4 05 47 PM" src="https://user-images.githubusercontent.com/72064462/232216517-ae853e29-c069-420c-9802-7b742895c527.png">


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

alya commented:
Thanks!

The spacing and sizing doesn't quite look right.

- For the `#`, we would probably be better off just using the `#` character without a space (`#announce`).
- The globe looks too big and probably too high.
- Both the globe an the lock need more space on the left.

akarsh-jain-790 commented:
> Thanks!
> 
> The spacing and sizing doesn't quite look right.
> 
> * For the `#`, we would probably be better off just using the `#` character without a space (`#announce`).
> * The globe looks too big and probably too high.
> * Both the globe an the lock need more space on the left.

@alya I have implemented all your suggested changes and also updated the PR description, please let me know if there is anything else I can improve!

alya commented:
Thanks! The icon sizes still don't quite look right to me (the lock seems significantly smaller than the others), but @amanagr 
perhaps you could review at this point? Is there another place in the code that could be used to match the icon size/alignment to in this PR?

amanagr commented on static/templates/bookend.hbs:
```diff
@@ -9,11 +9,23 @@
-                {{#tr}}You subscribed to stream {stream_name}{{/tr}}
+                {{#tr}}
+                    You subscribed to stream <z-stream-icon></z-stream-icon>
+                    {{#*inline "z-stream-icon"}}<i class="{{stream_icon}} stream-icon" aria-hidden="true"></i>{{/inline}}
+                    {{stream_name}}
+                {{/tr}}
             {{else if just_unsubscribed }}
-                {{#tr}}You unsubscribed from stream {stream_name}{{/tr}}
+                {{#tr}}
+                    You unsubscribed from stream <z-stream-icon></z-stream-icon>
+                    {{#*inline "z-stream-icon"}}<i class="{{stream_icon}} stream-icon" aria-hidden="true"></i>{{/inline}}
+                    {{stream_name}}
+                {{/tr}}
             {{else}}
-                {{#tr}}You are not subscribed to stream {stream_name}{{/tr}}
+                {{#tr}}
+                    You are not subscribed to stream <z-stream-icon></z-stream-icon>
+                    {{#*inline "z-stream-icon"}}<i class="{{stream_icon}} stream-icon" aria-hidden="true"></i>{{/inline}}
+                    {{stream_name}}
```
Can we use `stream_privacy.hbs` here?

akarsh-jain-790 replied:
@amanagr I'm not sure though I have updated the PR and description, you can re-review.

amanagr replied:
We should use `stream_privacy.hbs` here.

akarsh-jain-790 replied:
@amanagr I'm using the `stream_privacy.hbs`, and I noticed that the icon for the `#` is not displaying properly (as shown in the attached screenshot). It looks like the issue might be related to the fact that the CSS uses a span element for displaying the line, and the `#` icon in the `stream_privacy.hbs` file also uses a span element.

<img width="878" alt="Screenshot 2023-01-03 at 10 06 36 PM" src="https://user-images.githubusercontent.com/72064462/210404890-4d2b7055-e8fe-4d8f-b6f2-81ba7e5942b0.png">

amanagr replied:
Can you push that change in a different branch and link it here so that I can help you out?

akarsh-jain-790 replied:
Sure

https://github.com/akarsh-jain-790/zulip/tree/issue-23685-temp

amanagr replied:
@akarsh-jain-790 I don't understand what are you trying to do in that branch. I wanted you to use something like this:
```html
                {{#tr}}
                    You subscribed to stream {{> stream_privacy }}
                {{/tr}}
```

akarsh-jain-790 replied:
@amanagr Still getting the same result. It appears that the # icon is not displaying properly in the stream_privacy, while the globe and lock icons are functioning as expected. This discrepancy might be due to the conflicting use of span elements in both the CSS . 


> @amanagr I'm using the `stream_privacy.hbs`, and I noticed that the icon for the `#` is not displaying properly (as shown in the attached screenshot). It looks like the issue might be related to the fact that the CSS uses a span element for displaying the line, and the `#` icon in the `stream_privacy.hbs` file also uses a span element.
> 
> <img alt="Screenshot 2023-01-03 at 10 06 36 PM" width="878" src="https://user-images.githubusercontent.com/72064462/210404890-4d2b7055-e8fe-4d8f-b6f2-81ba7e5942b0.png">

<img width="892" alt="Screenshot 2023-02-07 at 12 18 02 PM" src="https://user-images.githubusercontent.com/72064462/217172435-48cda0ae-7445-4d05-862e-3986dfbab650.png">

<img width="872" alt="Screenshot 2023-02-07 at 12 38 33 PM" src="https://user-images.githubusercontent.com/72064462/217172632-20fb6f12-e492-479d-a9ce-cd2e3af0d6e3.png">

amanagr replied:
Ok

amanagr commented:
> The icon sizes still don't quite look right to me (the lock seems significantly smaller than the others).

Yes.

amanagr commented:
@akarsh-jain-790 can you match the size of icons and their alignment as they are in the recipient bar right now?

akarsh-jain-790 commented:
> @akarsh-jain-790 can you match the size of icons and their alignment as they are in the recipient bar right now? It would be helpful to review if you can post screenshots that show that they match.

@amanagr  would you kindly provide more specific details? I am not getting what you mean.

Joelute commented:
The recipient bar is the sticky bar we have up top:
![image](https://user-images.githubusercontent.com/62449508/219510886-c8549e6a-1f3d-44d8-a04a-aa1486e84ca7.png)

I'm pretty sure Aman wants you to match the icons and alignment to the ones in the recipient bar. Correct me if I'm wrong

akarsh-jain-790 commented:
@amanagr I have implemented your suggested changes and also updated the PR description, please let me know if there is anything else I can improve!

Joelute commented:
> It would be helpful to review if you can post screenshots that show that they match.

Screenshots would also be helpful! :)

akarsh-jain-790 commented:
> > It would be helpful to review if you can post screenshots that show that they match.
> 
> Screenshots would also be helpful! :)

@Joelute I've already updated the PR description to include the screenshots. Please take a look and let me know if there are any further changes you would like me to make.

alya commented:
The `#` still looks bigger than the others to me now.

amanagr commented:
Since we are using a new icon in [#23782](https://github.com/zulip/zulip/pull/23782), It might make sense to do this after it.

akarsh-jain-790 commented:
@alya @amanagr I've made the changes you suggested. Please take another look and let me know if you have any other feedback.

alya commented:
@akarsh-jain-790 , just to check, is this using the new icon from #23782 now?

akarsh-jain-790 commented:
> @akarsh-jain-790 , just to check, is this using the new icons from #23782 now?

Yes

alya commented:
Ok, great. The icons look good now, but the PR should be updated so that the privacy icons live-update when the privacy of a stream is changed. Works for me in manual testing other than that.

akarsh-jain-790 commented:
@alya Thank you for your feedback! I have updated the PR to include live-updating of the privacy icons when the privacy of a stream is changed.  Please let me know if there are any further changes or improvements you would like to see.

alya commented:
Thanks! I haven't re-tested, but maybe @amanagr you can take a look at this one next, when you have time?

amanagr commented:
LGTM! Thanks for the fix!

timabbott commented:
Merged, after a slight tweak to the templates to put the `{{#*inline "z-stream-icon"}}{{> stream_privacy }}{{/inline}}` line at the end of the block, since that's our usual convention and in my opinion a lot more readable. Thanks @akarsh-jain-790 and @amanagr!
